AsyncDisplayKit/Layout/ASLayout.mm

```diff
@@ -103,53 +127,61 @@ + (instancetype)flattenedLayoutWithLayoutableObject:(id<ASLayoutable>)layoutable
+                               sublayouts:layout.sublayouts];
 }
 
-- (ASLayout *)flattenedLayoutUsingPredicateBlock:(BOOL (^)(ASLayout *))predicateBlock
+#pragma mark - Layout Flattening
+
+- (ASLayout *)filteredNodeLayoutTree
 {
   NSMutableArray *flattenedSublayouts = [NSMutableArray array];
   
   struct Context {
     ASLayout *layout;
     CGPoint absolutePosition;
-    BOOL visited;
-    BOOL flattened;
   };
   
   // Queue used to keep track of sublayouts while traversing this layout in a BFS fashion.
   std::queue<Context> queue;
-  queue.push({self, CGPointMake(0, 0), NO, NO});
+  queue.push({self, CGPointMake(0, 0)});
   
   while (!queue.empty()) {
-    Context &context = queue.front();
-    if (context.visited) {
-      queue.pop();
-    } else {
-      context.visited = YES;
-      
-      if (predicateBlock(context.layout)) {
-        [flattenedSublayouts addObject:[ASLayout layoutWithLayoutableObject:context.layout.layoutableObject
-                                                       constrainedSizeRange:context.layout.constrainedSizeRange
-                                                                       size:context.layout.size
-                                                                   position:context.absolutePosition
-                                                                 sublayouts:nil
-                                                                  flattened:context.flattened]];
-      }
-      
-      for (ASLayout *sublayout in context.layout.sublayouts) {
-        // Mark layout trees that have already been flattened for future identification of immediate sublayouts
-        BOOL flattened = context.flattened ? : context.layout.flattened;
-        queue.push({sublayout, context.absolutePosition + sublayout.position, NO, flattened});
+    Context context = queue.front();
+    queue.pop();
+
+    if (self != context.layout && context.layout.type == ASLayoutableTypeDisplayNode) {
+      ASLayout *layout = [ASLayout layoutWithLayout:context.layout position:context.absolutePosition];
+      layout.flattened = YES;
+      [flattenedSublayouts addObject:layout];
+    }
+    
+    for (ASLayout *sublayout in context.layout.sublayouts) {
+      if (sublayout.isFlattened == NO) {
+        queue.push({sublayout, context.absolutePosition + sublayout.position});
       }
     }
   }
 
-  return [ASLayout flattenedLayoutWithLayoutableObject:_layoutableObject
-                                  constrainedSizeRange:_constrainedSizeRange
-                                                  size:_size
-                                            sublayouts:flattenedSublayouts];
+  return [ASLayout layoutWithLayoutableObject:_layoutableObject
+                         constrainedSizeRange:_constrainedSizeRange
+                                         size:_size
+                                   sublayouts:flattenedSublayouts];
+}
+
+#pragma mark - Accessors
+
+- (ASLayoutableType)type
```

Comment:
Do we need this accessor?  If so, should we call it layoutableType like the others?  @levi definitely like the idea of having this, as checking the class type had become necessary in some areas, including @hannahmbanana's layout inspector - so this helps cleanliness and efficiency.